src/core/service.c

```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
                 log_unit_debug(UNIT(s), "Passing %zu fds to service", exec_params.n_socket_fds + exec_params.n_storage_fds);
         }
 
+        /* These options require PrivateUsers= when used in user units, as we need to be in a user namespace
+         * to have permission to enable them when not running as root. */
+        if (!s->exec_context.private_users && !MANAGER_IS_SYSTEM(UNIT(s)->manager))
+                s->exec_context.private_users =
```

Comment:
please don't patch around in ExecContext here. we usually don't change loaded configuration, and expose the actually loaded stuff on the bus. if you'd change the setting here, then before the first time the the service is started we'll report one value here, and afterwards another.

instead in src/core/execute.c, look for the line:

```c
if (needs_sandboxing && context->private_users && !have_effective_cap(CAP_SYS_ADMIN)) {
```

and change that to also do the private user dance if the other options are set.

maybe add a helper accessor function:

```c
static bool exec_context_get_private_users(const ExecContext *c) {
       return c->private_users || c->n_bind_mounts > 0 || …;
}
```

and so on. and then change the aforementioned line to:

```c
if (needs_sandboxing && context_get_private_users(context) && !have_effective_cap(CAP_SYS_ADMIN)) {
```

you get the idea